server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GeoHashGridValuesSourceBuilder.java

```diff
@@ -0,0 +1,127 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.composite;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.geo.GeoUtils;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.geometry.utils.Geohash;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.bucket.geogrid.CellIdSource;
+import org.elasticsearch.search.aggregations.bucket.geogrid.GeoHashGridAggregationBuilder;
+import org.elasticsearch.search.aggregations.support.ValueType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class GeoHashGridValuesSourceBuilder extends CompositeValuesSourceBuilder<GeoHashGridValuesSourceBuilder> {
+    static final String TYPE = "geohash_grid";
+
+    private static final ObjectParser<GeoHashGridValuesSourceBuilder, Void> PARSER;
+    static {
+        PARSER = new ObjectParser<>(GeoHashGridValuesSourceBuilder.TYPE);
+        PARSER.declareInt(GeoHashGridValuesSourceBuilder::precision, new ParseField("precision"));
+        PARSER.declareInt(GeoHashGridValuesSourceBuilder::shardSize, new ParseField("shard_size"));
+        CompositeValuesSourceParserHelper.declareValuesSourceFields(PARSER, ValueType.NUMERIC);
+    }
+
+    static GeoHashGridValuesSourceBuilder parse(String name, XContentParser parser) throws IOException {
+        return PARSER.parse(parser, new GeoHashGridValuesSourceBuilder(name), null);
+    }
+
+    private int precision = GeoHashGridAggregationBuilder.DEFAULT_PRECISION;
+    private int shardSize = -1;
+
+    GeoHashGridValuesSourceBuilder(String name) {
+        super(name);
+    }
+
+    GeoHashGridValuesSourceBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.precision = in.readInt();
+        this.shardSize = in.readInt();
+    }
+
+    public GeoHashGridValuesSourceBuilder precision(int precision) {
+        this.precision = GeoUtils.checkPrecisionRange(precision);
+        return this;
+    }
+
+    public GeoHashGridValuesSourceBuilder shardSize(int shardSize) {
```

Comment:
What's the reason to add this option, is it really needed ?

Reply:
@jimczi I am not sure honestly. I was trying to keep parity with the `geohash_grid` aggregation and `shardSize` is used there. Maybe I misunderstand how composite aggs actually create and execute the underlying sources?

Reply:
I removed it, I read through and saw that composite does not utilize that parameter anywhere.